text/0000-thread-safe-handles.md

```diff
@@ -0,0 +1,362 @@
+
+As a developer, I may want to retain a reference to a JavaScript value while returning control back to the VM.
+
+`Handle<'_, _>` may not outlive the context that created them.
+
+```rust
+// Does not compile because `cb` cannot be sent across threads
+fn thread_log(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    let cb = cx.argument::<JsFunction>(0)?;
+
+    std::thread::spawn(move || {
+        println!("{:?}", cb);
+    });
+
+    cx.undefined()
+}
+```
+
+A persistent reference may be created from a `Handle<'_, _>` which can be sent across threads.
+
+```rust
+// Compiles
+fn thread_log(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    let cb = cx.argument::<JsFunction>(0)?.persistent(&mut cx);
+
+    std::thread::spawn(move || {
+        // Panics when `cb` is dropped.
+        println!("{:?}", cb);
+    });
+
+    cx.undefined()
+}
+```
+
+While `Persistent<_>` may be shared across threads, the inner contents can only be accessed on the main thread.
+
+```rust
+fn log(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    let persistent_cb = cx.argument::<JsFunction>(0)?.persistent(&mut cx);
+    let cb = persistent_cb.into_inner(&cx);
+
+    println!("{:?}", cb);
+
+    cx.undefined()
+}
+```
+
+`Persistent<_>` handles that are not unwrapped must be manually dropped. If a `Persistent<_>` is not consumed, neon will `panic` when the value is dropped.
+
+```rust
+fn log(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    let persistent_cb = cx.argument::<JsFunction>(0)?.persistent(&mut cx);
+
+    // `Persistent` are reference counted and clone-able.
+    let cb = persistent_cb.clone().into_inner(&cx);
+
+    // The callback is no longer needed
+    cb.drop(&cx);
+
+    cx.undefined()
+}
+```
+
+### `neon::handle::EventQueue`
+
+Once a value is wrapped in a `Persistent<_>` handle, it must be sent back to the main JavaScript thread to unwrap. `EventQueue` provides a mechanism for requesting work be peformed on the main thread.
+
+To schedule work, send a closure to the event queue:
+
+```rust
+fn thread_callback(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    let cb = cx.argument::<JsFunction>(0)?.persistent(&mut cx);
+    let queue = EventQueue::new(&mut cx)?;
+
+    std::thread::spawn(move || {
+        queue.send(move |mut cx| {
+            let this = cx.undefined();
+            let msg = cx.string("Hello, World!");
+            let cb = cb.into_inner(&cx);
+
+            cb.call(&mut cx, this, vec![msg]).unwrap();
+        });
+    });
+
+    cx.undefined()
+}
+```
+
+`Persistent<_>` are clone-able and can be used many times. `EventQueue` are `Sync` and may be shared across threads.
+
+```rust
+fn thread_callback(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    let cb = cx.argument::<JsFunction>(0)?.persistent(&mut cx);
+    let queue = Arc::new(EventQueue::new(&mut cx)?);
+
+    let handles = (1..=10).into_iter().map(|i| {
+        let queue = Arc::clone(queue);
```

Comment:
This particular line is giving me some heartburn. Something about it feels particularly difficult to learn and understand. Is there an ambiguity between the `EventQueue` and `Arc` implementations of `Clone` that requires the explicitly namespaced `Arc::clone()` instead of `queue.clone()`? Also why do you need the `Arc` in the first place if the `EventQueue` can be shared across multiple threads already? I'm just having trouble really explaining this part.

Reply:
`queue` does not implement `clone`. It *could*. This comes down to whether we want to utilize N-APIs underlying atomics on a function or leave it to the user.

The use of `Arc` is orthogonal to it being `Sync`. The `Arc` makes the lifetime `'static`, while `Sync` allows multiple threads to hold references. The `Arc` would be unnecessary here if the example used scoped threads.